### PR TITLE
Creates a new route to update archived status of users

### PR DESCRIPTION
**Pull request by Aryex82**
Issue: https://github.com/Real-Dev-Squad/website-backend/issues/1448

Description: 
This PR creates a new route which will run a script to sync the user roles - archived based on the users who are currently in RDS discord server.

- Should Do Nothing to Users With Archived as True and Not in RDS Discord Server
- Should Do Nothing to Users With Archived as False and in RDS Discord Server
- Should Archive Users With Archived as False and Not in RDS Discord Server
- Should Un-Archive Users With Archived as True and in RDS Discord Server

Note to reviewers: 
1. Unit tests for new functions in user model will be added in a separate PR.
2. API name might be changed

Screenshot: 
Screenshot 2023-09-02 at 12 33 05 AM

Test coverage:
Screenshot 2023-09-02 at 1 01 10 AM

### Related Issues
- #1474

### CODE REVIEW

File: controllers/external-accounts.js [hunk 1]
```diff
@@ -119,4 +120,94 @@ const syncExternalAccountData = async (req, res) => {
...
+          roles: {
+            ...rdsUser.roles,
+            in_discord: false,
+            archived: true,
+          },
+        };
+        updateUserList.push(userData);
+      }
+    }
+    const updateUserBatchPromise1 = updateUsersInBatch(updateUserList);
```
Comment (yesyash) [inline controllers/external-accounts.js:173]: - Should this be named `unArchieveUsersPromise` or something along those lines as we are just unarchiving users here?

File: controllers/external-accounts.js [hunk 2]
```diff
@@ -119,4 +120,94 @@ const syncExternalAccountData = async (req, res) => {
...
+        ...rdsUser,
+        roles: {
+          ...rdsUser.roles,
+          in_discord: true,
+          archived: false,
+        },
+      };
+      updateUserList.push(userData);
+    }
+    const updateUserBatchPromise2 = updateUsersInBatch(updateUserList);
```
Comment (yesyash) [inline controllers/external-accounts.js:191]: - Should this be named `archieveUsersPromise` or something along those lines as we are archiving users here?

File: constants/users.js [hunk 3]
```diff
@@ -19,6 +19,8 @@ const ALLOWED_FILTER_PARAMS = {
 
 const DOCUMENT_WRITE_SIZE = 500;
 
+const MAX_IN_OPERANDS = 30;
```
Comment (yesyash) [inline constants/users.js:22]: - What does `MAX_IN_OPERANDS` mean?

File: controllers/external-accounts.js [hunk 4]
```diff
@@ -119,4 +120,94 @@ const syncExternalAccountData = async (req, res) => {
   }
 };
 
-module.exports = { addExternalAccountData, getExternalAccountData, syncExternalAccountData };
+/**
+ * Gets all group-roles
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+const newSyncExternalAccountData = async (req, res) => {
```
[the remark this answers is not in the document]
Comment (Aryex82) [reply controllers/external-accounts.js:128]: yes, there's already a`SyncExternalAccountData` which will be deprecated with this

File: controllers/external-accounts.js [hunk 5]
```diff
@@ -119,4 +120,94 @@ const syncExternalAccountData = async (req, res) => {
...
+  try {
+    const [discordUserData, unArchivedRdsUsersData] = await Promise.all([
+      getDiscordMembers(),
+      fetchUsersForKeyValues("roles.archived", false),
+    ]);
+    let usersArchived = 0;
+    let usersUnArchived = 0;
+    let totalUsersProcessed = unArchivedRdsUsersData.length;
+
+    const discordUserIdSet = new Set();
```
[the remark this answers is not in the document]
Comment (Aryex82) [reply controllers/external-accounts.js]: - We are only using these values inside the try-block.
- Also `totalUsersProcessed` needs initial data from the api call. I think it makes more sense to declare all the count variable together.

File: controllers/external-accounts.js [hunk 6]
```diff
@@ -119,4 +120,94 @@ const syncExternalAccountData = async (req, res) => {
...
+ * @param res {Object} - Express response object
+ */
+const newSyncExternalAccountData = async (req, res) => {
+  try {
+    const [discordUserData, unArchivedRdsUsersData] = await Promise.all([
+      getDiscordMembers(),
+      fetchUsersForKeyValues("roles.archived", false),
+    ]);
+    let usersArchived = 0;
+    let usersUnArchived = 0;
```
[the remark this answers is not in the document]
Comment (Aryex82) [reply controllers/external-accounts.js]: Done

File: controllers/external-accounts.js [hunk 7]
```diff
@@ -119,4 +120,94 @@ const syncExternalAccountData = async (req, res) => {
...
+      if (discordUserIdSet.has(rdsUser?.discordId)) {
+        if (rdsUser.roles?.archived) usersUnArchived++;
+        userData = {
+          ...rdsUser,
+          roles: {
+            ...rdsUser.roles,
+            in_discord: true,
+            archived: false,
+          },
+        };
```
[the remark this answers is not in the document]
Comment (Aryex82) [reply controllers/external-accounts.js:160]: Yes, Thank you for pointing it out. Added a case where we can skip processing user data.

File: controllers/external-accounts.js (hunk 1 above)
Comment (Aryex82) [reply controllers/external-accounts.js]: Done.

File: controllers/external-accounts.js (hunk 2 above)
Comment (Aryex82) [reply controllers/external-accounts.js]: Done

File: constants/users.js (hunk 3 above)
Comment (Aryex82) [reply constants/users.js]: It's the maximum number of values allowed in firestore `IN` clause for querying data. Changed it to `FS_IN_CLAUSE_SIZE`. Let me know if it makes more sense now.

File: controllers/external-accounts.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (yesyash) [reply controllers/external-accounts.js:128]: - Can you please point me to the pr of ticket where we are depreciating this

File: constants/users.js (hunk 3 above)
Comment (yesyash) [reply constants/users.js]: - Can we removthe ambiguity around `FS` and name it  `FIRESTORE_IN_CLAUSE_SIZE` ?

File: controllers/external-accounts.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Aryex82) [reply controllers/external-accounts.js:128]: Its mentioned in the issue ticket: https://github.com/Real-Dev-Squad/website-backend/issues/1448|
- Will create a PR to remove the same once this API is in prod.

File: constants/users.js (hunk 3 above)
Comment (Aryex82) [reply constants/users.js]: done.

**Review by iamitprakash — APPROVED** [review]

File: controllers/external-accounts.js [hunk 8]
```diff
@@ -119,4 +120,95 @@ const syncExternalAccountData = async (req, res) => {
...
+    const discordUserIdSet = new Set();
+
+    discordUserData.forEach((discordUser) => discordUserIdSet.add(discordUser.user.id));
+
+    let updateUserList = [];
+
+    for (const rdsUser of unArchivedRdsUsersData) {
+      let userData = {};
+
+      // This if-block will be removed if the IN_DISCORD ROLE is deprecated.
```
Comment (ankushdharkar) [inline controllers/external-accounts.js:147]: Future-dated ticket please with a TODO comment

File: controllers/external-accounts.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Aryex82) [reply controllers/external-accounts.js:128]: Issue to track deprecating the old API: https://github.com/Real-Dev-Squad/website-backend/issues/1474

File: controllers/external-accounts.js (hunk 8 above)
Comment (Aryex82) [reply controllers/external-accounts.js]: Made the necessary changes.

File: controllers/external-accounts.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (yesyash) [reply controllers/external-accounts.js:128]: - Added #1474 to the PR description.

**Review by yesyash — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]
LGTM

**Pull request merged**